### PR TITLE
OpenMapTiles 3.15.0 SNAPSHOT (3)

### PR DESCRIPTION
**Pull request by phanecak-maptiler**
Follow-up on #144 , contains bugfixes for problems found while comparing tilesets generated with 3.15.0 SNAPSHOT of OpenMapTiles and `planetiler-openmaptiles` and also final bump of version from `3.14.0` to `3.15.0`:

- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/41
  - plus https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/42
  - plus https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/50
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/43
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/44
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/45
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/46
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/47
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/48
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/49
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/51
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/55
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/58

### CODE REVIEW

**Comment by phanecak-maptiler** [comment]
note: Once this is merged, it fixes #15 thanks to https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/48 .

**Review by lazaa32 — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -165,8 +166,8 @@ If you want to regenerate from a different repository than the default openmapti
 All code in this repository is under the [BSD license](./LICENSE.md) and the cartography decisions encoded in the schema
 and SQL are licensed under [CC-BY](./LICENSE.md).
 
-Products or services using maps derived from OpenMapTiles schema need to visibly credit "OpenMapTiles.org" or
-reference "OpenMapTiles" with a link to https://openmaptiles.org/. Exceptions to attribution requirement can be granted
+Products or services using maps derived from OpenMapTiles schema need to **visibly credit "OpenMapTiles.org"** or
+**reference "OpenMapTiles"** with a link to https://openmaptiles.org/. Exceptions to attribution requirement can be granted
```
Comment (msbarry) [inline README.md:170]: Interesting to add the emphasis here, I thought the plan was to switch to CC0?

File: src/main/java/org/openmaptiles/layers/WaterName.java [hunk 2]
```diff
@@ -214,38 +218,63 @@ public void process(Tables.OsmMarinePoint element, FeatureCollector features) {
...
+        if ("bay".equals(element.natural())) {
+          clazz = FieldValues.CLASS_BAY;
+        } else if ("sea".equals(place)) {
+          clazz = FieldValues.CLASS_SEA;
+        } else {
+          clazz = FieldValues.CLASS_LAKE;
+          minzoomCL = MINZOOM_LAKE;
+        }
+        if (centerlineGeometry != null) {
+          // prefer lake centerline if it exists, but point will be also used if minzoom bellow 9 is calculated from area
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:235]: typ: `bellow` -> `below`

File: src/main/java/org/openmaptiles/layers/WaterName.java [hunk 3]
```diff
@@ -214,38 +218,63 @@ public void process(Tables.OsmMarinePoint element, FeatureCollector features) {
...
+          // note: Here we're diverging from OpenMapTiles: For bays with minzoom (based on area) point is used between
+          // minzoom and Z8 and for Z9+ centerline is used, while OpenMaptiles sticks with points.
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:237]: why? Is planetiler-openmaptiles handling considered "better" here?

File: src/main/java/org/openmaptiles/layers/WaterName.java [hunk 4]
```diff
@@ -214,38 +218,63 @@ public void process(Tables.OsmMarinePoint element, FeatureCollector features) {
...
+        }
+
+        // Show a label if a water feature covers at least 1/4 of a tile or z14+
+        Geometry geometry = element.source().worldGeometry();
+        double area = geometry.getArea();
+        int minzoom = (int) Math.floor(-1d - Math.log(Math.sqrt(area)) / LOG2);
+        if (place != null && SEA_OR_OCEAN_PLACE.contains(place)) {
+          minzoom = Math.clamp(minzoom, MINZOOM_SEA_AND_OCEAN, 14);
+        } else {
+          minzoom = Math.clamp(minzoom, MINZOOM_LAKE, 14);
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:250]: You should be able to replace this handling with a combination of min area and min zoom?

```java
feature.setMinPixelSizeBelowZoom(13, 128).setMinZoom(minzoom)
```

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 5]
```diff
@@ -121,11 +177,135 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+        intersection = neGeom.intersection(fixedGeom);
+      } catch (TopologyException e2) {
+        throw new GeometryException("fix_omt_water_topology_error",
+          "error fixing polygon: " + e2 + "; original error: " + e);
+      }
+    }
+
+    // should match following in OpenMapTiles: Distinct on keeps just the first occurence -> order by 'area_ratio DESC'
+    double areaRatio = intersection.getArea() / neGeom.getArea();
+    if (areaRatio > lakeInfo.areaRatio) {
```
[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Water.java:254]: Good point. I did it the same way as OpenMaptiles, but since "bigger area" implies "bigger ratio" we can drop `/ neGeom.getArea()`. Adjusted, thanks.

File: src/main/java/org/openmaptiles/layers/WaterName.java (hunk 2 above)
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java:235]: Adjusted.

File: src/main/java/org/openmaptiles/layers/WaterName.java (hunk 3 above)
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java:237]: Bay name is rendered along the centerline, instead of as usual label used for points.

File: src/main/java/org/openmaptiles/layers/WaterName.java (hunk 4 above)
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java:250]: Will that work if we're creating a point from that polygon with `features.pointOnSurface()`?

Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/WaterName.java]: Yes! I updated it to work that way in order to support use-cases like this here: https://github.com/onthegomap/planetiler/pull/720

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 6]
```diff
@@ -86,21 +105,56 @@ record WaterInfo(int minZoom, int maxZoom, String clazz) {}
...
+        synchronized (this) {
+          neAllLakeInfos.add(lakeInfo);
+        }
+        neLakeIndex.put(geom, lakeInfo);
+        if (lakeInfo.name != null) {
+          if (!neLakeNameMap.containsKey(lakeInfo.name) ||
+            lakeInfo.geom.getArea() > neLakeNameMap.get(lakeInfo.name).geom.getArea()) {
+            // on name collision, bigger lake gets on the name list
+            neLakeNameMap.put(lakeInfo.name, lakeInfo);
+          }
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:139]: I think we want to do an atomic operation on the concurrent hashmap here to avoid multiple updates at the same time from colliding with eachother:

```java
neLakeNameMap.merge(lakeInfo.name, (prev, next) -> next.geom.getArea() > prev.geom.getArea() ? next : prev);
```

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 7]
```diff
@@ -121,11 +175,123 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+      lakeInfo.area = area;
+    }
+  }
+
+  @Override
+  public void finish(String sourceName, FeatureCollector.Factory featureCollectors,
+    Consumer<FeatureCollector.Feature> emit) {
+    if (OpenMapTilesProfile.NATURAL_EARTH_SOURCE.equals(sourceName)) {
+      var timer = stats.startStage("ne_lake_index");
+      timer.stop();
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:254]: do you still need this?

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 8]
```diff
@@ -121,11 +175,123 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+    var intersection = neGeom.intersection(geom);
+
+    // Should match following in OpenMapTiles: Distinct on keeps just the first occurence -> order by 'area_ratio DESC'
+    // With a twist: NE geometry is always the same, hence we can make it a little bit faster by dropping "ratio"
+    // and compare only the intersection area: bigger area -> bigger ratio.
+    double area = intersection.getArea();
+    if (area > lakeInfo.area) {
+      lakeInfo.osmId = element.source().id();
+      lakeInfo.area = area;
+    }
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:246]: seems like there might be a risk here of multiple OSM threads updating the same lakeInfo instance at the same time, would probably be safer to wrap this in `synchronized (lakeInfo) { ... }`

File: src/main/java/org/openmaptiles/OpenMapTilesMain.java [hunk 9]
```diff
@@ -4,12 +4,16 @@
 import com.onthegomap.planetiler.config.Arguments;
```
[the remark this answers is not in the document]
Comment (msbarry) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: OK I see you fixed the roads at z4/5, but I'm still seeing these discrepancies in planetiler-openmaptiles:

- M74 doesn't have "e-road" network set: https://onthegomap.github.io/planetiler-demo/#17/55.687717/-3.942389 - did we decide to completely get rid of e-roads and AsianHighway designation after that omt pr?
- That Peru maritime boundary still shows up in planetiler-openmaptiles output at z4: https://onthegomap.github.io/planetiler-demo/#4.4/-9.01/-78.94

Those are showing latest output from main branch of planetiler-openmaptiles from @ZeLonewolf's setup

**Review by msbarry — COMMENTED** [review]
looks good!  a few more comments...

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 10]
```diff
@@ -121,11 +175,123 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+        var geom = element.source().worldGeometry();
+        try {
+          attemptNeLakeIdMapping(element, geom);
+        } catch (TopologyException e) {
+          try {
+            var fixedGeom = GeometryFixer.fix(geom);
+            attemptNeLakeIdMapping(element, fixedGeom);
+          } catch (TopologyException e2) {
+            throw new GeometryException("fix_omt_water_topology_error",
+              "error fixing polygon: " + e2 + "; original error: " + e);
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:189]: I see this is firing 16 times when run over the planet, but I can't tell which line is actually causing this failure?  Can you catch the TopologyException closer to where it happens so that 1) we get a more detailed error message that helps pinpoint where the issue happened and 2) we can skip the outer layer of try/catch(geometry exception) and just have the final attempt log the error like this: https://github.com/onthegomap/planetiler/blob/12558b99a8074781e2adc9a2130cab5f481b4395/planetiler-core/src/main/java/com/onthegomap/planetiler/geo/GeoUtils.java#L311-L338.  Also please pass `config.logJtsExceptions()` into the 4th argument of `e.log(...)` below which will print out more details about when the error happens?

I understand this is a best-effort attempt to add OSM IDs to low zoom lakes so it's fine if it misses some, we just try to track down any of these robustness issues with JTS pretty aggressively to report to its maintainer.

Here's an example failure for reference:

> 0:56:26 WAR [osm_pass2:process] - Unable to process intersections for OSM feature 13070460: error fixing polygon: org.locationtech.jts.geom.TopologyException: found non-noded intersection between LINESTRING ( 0.7828271086875376 0.3132446324915123, 0.7828058490561336 0.3132069744583109 ) and LINESTRING ( 0.7828058490561336 0.3132069744583109, 0.7828042136998719 0.3132040773818683 ) [(0.7828058490561336, 0.3132069744583109, NaN) ]; original error: org.locationtech.jts.geom.TopologyException: found non-noded intersection between LINESTRING ( 0.7828271086875376 0.3132446324915123, 0.7828058490561336 0.3132069744583109 ) and LINESTRING ( 0.7828058490561336 0.3132069744583109, 0.7828042136998719 0.3132040773818683 ) [ (0.7828058490561336, 0.3132069744583109, NaN) ]

File: src/main/java/org/openmaptiles/OpenMapTilesMain.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (msbarry) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: Looks like this is still an issue on this branch:
- e-road: https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F%7Emsbarry%2Fomt-3.15-branch.pmtiles#map=17/55.687717/-3.942389
- Peru border: https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F%7Emsbarry%2Fomt-3.15-branch.pmtiles#map=4.4/-9.01/-78.94

File: src/main/java/org/openmaptiles/layers/Transportation.java [hunk 11]
```diff
@@ -245,6 +255,40 @@ private static boolean isResidentialOrUnclassified(String highway) {
...
+    if (!"trunk".equals(highway)) {
+      return false;
+    }
+    return routeRelations.stream()
+      .map(RouteRelation::networkType)
+      .filter(Objects::nonNull)
+      .anyMatch(Z5_TRUNK_BY_NETWORK::contains);
+  }
+
+  private static boolean isMotorwayWithNetworkForZ4(List<RouteRelation> routeRelations) {
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:269]: should this be named `isTrunkWithNetworkForZ4` ? Or should we also run this check down below:

```java
        case FieldValues.CLASS_MOTORWAY ->
          isMotorwayForZ4(routeRelations) ||
          isMotorwayWithNetworkForZ4(routeRelations) ?
          MINZOOMS.getOrDefault(FieldValues.CLASS_MOTORWAY, Integer.MAX_VALUE) : 5;
```

File: src/main/java/org/openmaptiles/layers/Transportation.java [hunk 12]
```diff
@@ -245,6 +255,40 @@ private static boolean isResidentialOrUnclassified(String highway) {
...
+      .filter(nt -> !Z5_MOTORWAYS_BY_NETWORK.contains(nt))
+      .anyMatch(Z5_TRUNK_BY_NETWORK::contains);
+  }
+
+  private static boolean isMotorwayWoNetworkForZ4(List<RouteRelation> routeRelations) {
+    // All motorways without network (e.g. EU, Asia, South America)
+    return routeRelations.stream()
+      .map(RouteRelation::networkType)
+      .noneMatch(Objects::nonNull);
+  }
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:283]: I notice that this logic leads to short stretches of road in the US getting included at z4 because they have class=motorway without a route network and they are disconnected with the rest of the road because we exclude class=motorway/network=us-highway until z5 - is that the intent? 

I think it's because this logic is upgrading those trunks to motorways:

```java
          if (z5trunk && isMotorwayWithNetworkForZ4(routeRelations)) {
            clazz = FieldValues.CLASS_MOTORWAY;
```

image


https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F%7Emsbarry%2Fomt-3.15-branch.pmtiles#map=4.8/36.05/-84.16

File: src/main/java/org/openmaptiles/layers/Transportation.java [hunk 13]
```diff
@@ -296,11 +340,7 @@ public List<OsmRelationInfo> preprocessOsmRelation(OsmElement.Relation relation)
       String colour = coalesce(
         nullIfEmpty(relation.getString("colour")), nullIfEmpty(relation.getString("ref:colour")));
 
-      if ("e-road".equals(network)) {
-        networkType = RouteNetwork.E_ROAD;
-      } else if ("AsianHighway".equals(network)) {
-        networkType = RouteNetwork.A_ROAD;
-      } else if ("US:I".equals(network)) {
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:303]: did we mean to remove these? If so we should remove them from the enum list or otherwise keep them here.

File: src/main/java/org/openmaptiles/OpenMapTilesMain.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: As per M74: That seems to be related to https://github.com/openmaptiles/openmaptiles/pull/1648 and https://github.com/openmaptiles/openmaptiles/pull/1649 . I've regenerated that portion of UK and `e-road` is there:

Screenshot from 2024-04-29 13-47-00

Hence I guess Brian's tiles were done with older snapshot.

[the remark this answers is not in the document]
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: As per Peru: Fix/change is here, originally tracked here: https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/47 . Generated with current snapshot:

Screenshot from 2024-04-29 13-53-08

So again, maybe older snapshot used by Brian.

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 14]
```diff
@@ -121,11 +182,147 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+    final Geometry neGeom = lakeInfo.geom;
+    Geometry intersection;
+    try {
+      if (intersetsCheckNeeded && !neGeom.intersects(geom)) {
+        return;
+      }
+      intersection = neGeom.intersection(geom);
+    } catch (TopologyException e) {
+      stats.dataError("omt_water_intersection");
+      LOGGER.debug("omt_water_intersection, NE ID: {}, OSM ID: {}",
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:253]: Maybe add a little bit more context, like "fixed invalid OSM water polygon {} when adding ID to natural earth {}"

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 15]
```diff
@@ -121,11 +182,147 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+    }
+
+    // match by intersection:
+    List<LakeInfo> items;
+    try {
+      items = neLakeIndex.getIntersecting(geom);
+    } catch (TopologyException e) {
+      stats.dataError("omt_water_intersecting");
+      LOGGER.debug("omt_water_intersecting, OSM ID: {}",
+        element.source().id(), e);
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:224]: Maybe add a little bit more context, like "fixed invalid OSM water polygon {} when looking up natural earth features"

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 16]
```diff
@@ -121,11 +182,147 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+      final var geomFixed = GeometryFixer.fix(geom);
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:225]: can you overwrite geom here so that fillOsmIdIntoNeLake doesn't need to fix the geometry again?

**Review by msbarry — COMMENTED** [review]
looks good, a few tweaks make the intersection error logs a bit more self-explanatory...

File: src/main/java/org/openmaptiles/layers/Water.java (hunk 15 above)
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:224]: also let's omit the exception here and on line 253 below - there's no sense digging into JTS exceptions unless we're sure the input is valid.

**Comment by phanecak-maptiler** [comment]
Regarding latest batch of "adjusted" comments from me: You can see that only now, since only now I've pushed and merged those changes. Sorry for the delay and confusion.

File: src/main/java/org/openmaptiles/layers/Landcover.java [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2021, MapTiler.com & OpenMapTiles contributors.
+Copyright (c) 2023, MapTiler.com & OpenMapTiles contributors.
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Landcover.java:2]: I notice you've got 2023 in a few of these headers

File: src/main/java/org/openmaptiles/layers/Water.java [hunk 18]
```diff
@@ -121,11 +182,122 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
...
+    // if OSM lake is too small for Z6 (e.g. area bellow ~4px) we assume there is no matching NE lake
+    var geom = element.source().worldGeometry();
+    if (geom.getArea() < OSM_ID_MATCH_AREA_LIMIT) {
+      return;
+    }
+
+    if (!geom.isValid()) {
+      geom = GeometryFixer.fix(geom);
+      stats.dataError("omt_fix_water_before_ne_intersect");
+      LOGGER.debug("Fixing geometry of OSM element {} before attempt to add ID to natural earth water feature", element.source().id());
```
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:205]: looks like you need to run `./scripts/format.sh`

**Review by msbarry — APPROVED** [review]
looks good!

File: src/main/java/org/openmaptiles/layers/Landcover.java (hunk 17 above)
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Landcover.java:2]: Ah, @TomPohys was asking about the same.

TL;DR: IANAL hence I tend to put there a year of "most recent substantial change". This particular file has such in 2023.

Longer: Looking through various IP advice and Open Source licensing FAQs, etc. there seems to be varying degree of advice, from "does not matter" to "really important", with various tactics about when and how to change that value. And since I'm not going to hire a layer (which is IMHO important if we truly need a legal advice about for example what "substantial change" means), then I outline my rationale:

1. Extreme position 1: Bump the copyright year every year even when no changes were done = IMHO totally bogus and IANAL but legally unenforceable.
2. Extreme position 1: Put there a year at the beginning and then never bump it, even if the file goes through substantial changes all the time = IMHO also wrong, but sort of "wronging myself" (e.g. possibly loosing protections which I'm entitled to, but hey, copyright protection is supposed to run even several years after my death, so ...).

Hence I opt to be "somewhere in between", e.g. bump the year when substantial changes occur.

File: src/main/java/org/openmaptiles/layers/Water.java (hunk 18 above)
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Water.java]: Thanks, adjusted.

**Comment by phanecak-maptiler** [comment]
Merged.

Release planned for Monday. E.g. if no serious showstopper is found until than, this will be 3.15.
